### ISSUE TITLE
Why is conforming to ListBindable required?

### ISSUE THREAD

**Issue by arctouch-marcelogobetti** [Instagram/IGListKit#893]
## New issue checklist

- [x] I have reviewed the [`README`](https://github.com/Instagram/IGListKit/blob/master/README.md) and [documentation](http://instagram.github.io/IGListKit)
- [x] I have searched [existing issues](https://github.com/Instagram/IGListKit/issues) and this is not a duplicate

### General information

- `IGListKit` version: 3.0
- iOS version(s): Any
- CocoaPods/Carthage version: 1.2.1
- Xcode version: 8.3.3
- Devices/Simulators affected: Any
- Reproducible in the demo project? (Yes/No): Yes
- Related issues: #449

### Debug information

Not attaching debug information since this is just a question.

### Question

First of all, thank you so much for this binding section controller - I was frustrated about how cumbersome it would be to dynamically update certain parts in our app, and suddenly in a few hours that was gracefully handled by another IGListKit component :)

The only part I didn't like (yet!) though was that I was required to implement the `ListBindable` protocol. I know: I'm probably doing it wrong, but because of the nature of my project I don't see any issues from what I'm doing (I don't ever update existent cells, I do only append more).

When using the regular ListSectionController class, I was creating and setting up my cells from the `cellForItem:` method. Now we have `cellForViewModel:atIndex:` - ok, I've just moved the code over to the new method basically. We're not even using the view model parameter, to be honest, given that the section controller owns an array of objects and thus it knows exactly which object to take from the array (by its index). I can still see my `viewModelsFor:` being useful - internally - since they're input to `IGListDiff`.

All good, let's run the project - crash because our cells do not implement `ListBindable`. So checking your `LabelCell` example, this method's responsibility is basically hooking a model to the view (couldn't have guessed!). So let's say my cells are also `LabelCell`s: I'm already doing `label.text = objects[index] as? String` in `cellForViewModel:atIndex:`, and I really don't see (yet!) a reason to review my architecture in tens of cells just to separate this line into a new method. I mean, I can see why it looks good to have this separated, such as being able to keep things private in the cell, but it has never been *necessary* on iOS nor on IGListKit, and trust me: we've really done a big architecture with Swift protocols that allows us to export this `cellForItem:` logic to outer classes, it looks and works good and the question is exactly: why would we need to change?

Given the limitations of adding default implementations to `@objc` protocols, we've added the boilerplate `ListBindable` conformance with empty `bindViewModel:` implementation to every cell. It just works. Even being empty. In `IGListBindingSectionController.m`, I see a cell being built at an `oldUpdatedIndex` and then having a model from an `indexAfterUpdate` hook up to it, so that's kind of a big hint already, except that my view models indices do never update. I'd be convinced that my scenario is just too specific, yet apparently [we're not the only ones doing it](https://github.com/facebookarchive/AsyncDisplayKit/blob/master/Source/Details/_ASCollectionViewCell.m) and after applying the same reasoning to your `MonthSectionController` example (moved `ListBindable` implementation from `LabelCell` into `cellForItem:` in the section controller), it still works exactly as before.

### TL;DR

1) Is it possible to make `ListBindable` conformance optional?
2) If not, what are exactly all the functional issues when not separating code out from `cellForItem:` into `bindViewModel:`?

Current code:
```
UICollectionViewCell<IGListBindable> *cell = [collectionContext cellForItemAtIndex:oldUpdatedIndex
                                                                 sectionController:self];
[cell bindViewModel:self.viewModels[indexAfterUpdate]];
```

What I propose:
```
UICollectionViewCell<IGListBindable> *cell = [collectionContext cellForItemAtIndex:oldUpdatedIndex
                                                                 sectionController:self];
if ([cell conformsToProtocol:@protocol(IGListBindable)]) {
    [cell bindViewModel:self.viewModels[indexAfterUpdate]];
}
```

Thank you so much!

**Comment by rnystrom** [comment]
Great questions! And you did quite a lot of investigation too.

You can absolutely still do your cell config in `cellForItemAtIndex:`. It's exactly the same as doing it in the bind method. 

The binding exists because of how common practice it is to configure a cell given a view model when the model drives how many and what type of cells exist. And you're right, it helps separate code.

But if you have your own architecture, that's good too!

Also, sometimes you will still have to do extra work in `cellForItemAtIndex:`, e.g. If your cell has a delegate.

Why is it required? For perf reasons. Checking protocol conformance is immensely more expensive than checking for class type due too ObjCs dynamic nature with protocols and optionality. This code executes in Instagram, so we want it lightning fast.

Empty bind methods are totally fine (and MUCH faster).

**Comment by arctouch-marcelogobetti** [comment]
Great, I confess that's news to me :)
I'm usually all for avoiding boilerplate - in which case I'd probably rather not having `ListBindable` at all - but as I said, I know it is an architecture improvement, so it's fine. Thanks a lot!
